File: cmd/conformance/normalize.go
package main

import (
	"fmt"
	"path/filepath"
	"sort"

	"google.golang.org/protobuf/reflect/protoreflect"
)

// Placeholders a normalized response carries in place of a value that is not
// the same twice. Scenarios compare against these spellings.
const (
	modelHashPlaceholder = "${model_hash}"
	versionPlaceholder   = "${version}"
	pathPlaceholder      = "${path}"
)

// normalizedIDs are the int64 fields carrying a runtime instance id, which is
// assigned per call: each distinct id becomes "@1", "@2", … in order of first
// appearance, so a scenario can state that two feature values name one object
// without knowing which id it was given.
var normalizedIDs = map[string]bool{
	"sysml.Instance.id":         true,
	"sysml.Value.instance_id":   true,
	"sysml.Verdict.instance_id": true,
	"sysml.Function.self_id":    true,
}

// integer and unsigned are normalized integral values. They are distinct from
// float64 so an integral field is compared exactly rather than within the
// tolerance a Real carries.
type (
	integer  int64
	unsigned uint64
)

// normalizer turns a response into a JSON-shaped tree with the values that
// cannot be compared literally replaced. See conformance/README.md.
type normalizer struct {
	modelHash string
	labels    map[int64]string
}

func newNormalizer(modelHash string) *normalizer {
	return &normalizer{modelHash: modelHash, labels: map[int64]string{}}
}

// normalize renders msg as maps, slices, strings, numbers and bools. Only set
// fields appear; a scalar left at its default is absent.
func (n *normalizer) normalize(msg protoreflect.Message) map[string]any {
	out := map[string]any{}
	var set []protoreflect.FieldDescriptor
	msg.Range(func(field protoreflect.FieldDescriptor, _ protoreflect.Value) bool {
		set = append(set, field)
		return true
	})
	// Range yields fields in no defined order; number order makes id labels
	// reproducible.
	sort.Slice(set, func(i, j int) bool { return set[i].Number() < set[j].Number() })

	for _, field := range set {
		value := msg.Get(field)
		switch {
		case field.IsMap():
			entries := map[string]any{}
			keys := make([]string, 0, value.Map().Len())
			byKey := map[string]protoreflect.Value{}
			value.Map().Range(func(key protoreflect.MapKey, item protoreflect.Value) bool {
				keys = append(keys, key.String())
				byKey[key.String()] = item
				return true
			})
			sort.Strings(keys)
			for _, key := range keys {
				entries[key] = n.value(field.MapValue(), byKey[key])
			}
			out[field.TextName()] = entries
		case field.IsList():
			list := value.List()
			items := make([]any, 0, list.Len())
			for i := 0; i < list.Len(); i++ {
				items = append(items, n.value(field, list.Get(i)))
			}
			out[field.TextName()] = items
		default:
			out[field.TextName()] = n.value(field, value)
		}
	}
	return out
}

// value normalizes one field value, list element or map value.
func (n *normalizer) value(field protoreflect.FieldDescriptor, value protoreflect.Value) any {
	full := string(field.FullName())
	switch field.Kind() {
	case protoreflect.MessageKind, protoreflect.GroupKind:
		return n.normalize(value.Message())
	case protoreflect.EnumKind:
		return string(field.Enum().Values().ByNumber(value.Enum()).Name())
	case protoreflect.BoolKind:
		return value.Bool()
	case protoreflect.StringKind:
		return n.stringValue(full, value.String())
	case protoreflect.DoubleKind, protoreflect.FloatKind:
		return value.Float()
	case protoreflect.Int64Kind, protoreflect.Sint64Kind, protoreflect.Sfixed64Kind:
		if normalizedIDs[full] {
			return n.label(value.Int())
		}
		return integer(value.Int())
	case protoreflect.Int32Kind, protoreflect.Sint32Kind, protoreflect.Sfixed32Kind:
		return integer(value.Int())
	case protoreflect.Uint32Kind, protoreflect.Uint64Kind, protoreflect.Fixed32Kind, protoreflect.Fixed64Kind:
		return unsigned(value.Uint())
	default:
		return value.String()
	}
}

// stringValue replaces the strings a call cannot repeat: the model hash it was
// given, the service's build version, and any absolute path.
func (n *normalizer) stringValue(field, text string) string {
	switch {
	case field == "sysml.ServerInfoResponse.version":
		return versionPlaceholder
	case n.modelHash != "" && text == n.modelHash:
		return modelHashPlaceholder
	case filepath.IsAbs(text):
		return pathPlaceholder
	default:
		return text
	}
}

// label is the symbolic name of a runtime instance id.
func (n *normalizer) label(id int64) string {
	if existing, ok := n.labels[id]; ok {
		return existing
	}
	label := fmt.Sprintf("@%d", len(n.labels)+1)
	n.labels[id] = label
	return label
}

File: clients/node/conformance/normalize.ts
// Turning a response into the JSON-shaped tree scenarios compare against, with
// the values that are not the same twice replaced. See conformance/README.md.

import { ScalarType, type DescField, type DescMessage, type Message } from "@bufbuild/protobuf";
import { isAbsolute } from "node:path";
import { reflect, type ReflectMessage } from "@bufbuild/protobuf/reflect";

import { byCodeUnit } from "./order.js";

export const MODEL_HASH_PLACEHOLDER = "${model_hash}";
export const VERSION_PLACEHOLDER = "${version}";
export const PATH_PLACEHOLDER = "${path}";

/** The int64 fields carrying a runtime instance id, relabelled per call. */
const NORMALIZED_IDS = new Set([
  "sysml.Instance.id",
  "sysml.Value.instance_id",
  "sysml.Verdict.instance_id",
  "sysml.Function.self_id",
]);

/**
 * An integral value, kept as its digits. Distinct from a number so an integral
 * field is compared exactly rather than within the tolerance a Real carries.
 */
export class Integer {
  constructor(readonly digits: string) {}

  toString(): string {
    return this.digits;
  }
}

/** A normalized response: maps, lists, strings, numbers, Integers and booleans. */
export type Normalized = Record<string, unknown>;

/** Replaces the values a call cannot repeat, and labels its instance ids. */
export class Normalizer {
  private readonly labels = new Map<string, string>();

  constructor(private readonly modelHash: string) {}

  /** Renders a message as a tree. Only set fields appear. */
  normalize(schema: DescMessage, message: Message): Normalized {
    return this.message(reflect(schema, message));
  }

  private message(reflected: ReflectMessage): Normalized {
    const out: Normalized = {};
    // Number order, so instance id labels are assigned reproducibly.
    const fields = [...reflected.fields].sort((left, right) => left.number - right.number);
    for (const field of fields) {
      if (!reflected.isSet(field)) {
        continue;
      }
      out[field.name] = this.field(field, reflected);
    }
    return out;
  }

  private field(field: DescField, reflected: ReflectMessage): unknown {
    switch (field.fieldKind) {
      case "list": {
        const list = reflected.get(field);
        return [...list].map((item) => this.element(field, item));
      }
      case "map": {
        const map = reflected.get(field);
        const entries: Normalized = {};
        const keys = [...map.keys()].map(String).sort(byCodeUnit);
        for (const key of keys) {
          entries[key] = this.element(field, map.get(key));
        }
        return entries;
      }
      case "message":
        return this.message(reflected.get(field));
      case "enum":
        return this.enumName(field, reflected.get(field));
      case "scalar":
        return this.scalar(field, reflected.get(field));
    }
  }

  /** Normalizes one list element or map value, whose kind the field declares. */
  private element(field: DescField, value: unknown): unknown {
    let kind: "message" | "enum" | "scalar" | undefined;
    if (field.fieldKind === "list") {
      kind = field.listKind;
    } else if (field.fieldKind === "map") {
      kind = field.mapKind;
    }
    switch (kind) {
      case "message":
        return this.message(value as ReflectMessage);
      case "enum":
        return this.enumName(field, value);
      default:
        return this.scalar(field, value);
    }
  }

  private enumName(field: DescField, value: unknown): string {
    const descriptor = field.enum;
    const number = Number(value);
    const literal = descriptor?.values.find((candidate) => candidate.number === number);
    return literal?.name ?? String(number);
  }

  private scalar(field: DescField, value: unknown): unknown {
    const scalar = field.scalar ?? (field.fieldKind === "map" ? field.scalar : undefined);
    const full = `${field.parent.typeName}.${field.name}`;
    switch (scalar) {
      case ScalarType.DOUBLE:
      case ScalarType.FLOAT:
        return Number(value);
      case ScalarType.INT64:
      case ScalarType.SINT64:
      case ScalarType.SFIXED64:
        return NORMALIZED_IDS.has(full) ? this.label(String(value)) : new Integer(String(value));
      case ScalarType.INT32:
      case ScalarType.SINT32:
      case ScalarType.SFIXED32:
      case ScalarType.UINT32:
      case ScalarType.UINT64:
      case ScalarType.FIXED32:
      case ScalarType.FIXED64:
        return new Integer(String(value));
      case ScalarType.BOOL:
        return Boolean(value);
      case ScalarType.BYTES:
        return Buffer.from(value as Uint8Array).toString("base64");
      default:
        return this.string(full, String(value));
    }
  }

  /** Replaces the model hash it was given, the build version, and absolute paths. */
  private string(field: string, text: string): string {
    if (field === "sysml.ServerInfoResponse.version") {
      return VERSION_PLACEHOLDER;
    }
    if (this.modelHash !== "" && text === this.modelHash) {
      return MODEL_HASH_PLACEHOLDER;
    }
    if (isAbsolute(text)) {
      return PATH_PLACEHOLDER;
    }
    return text;
  }

  /** The symbolic name of a runtime instance id. */
  private label(id: string): string {
    const existing = this.labels.get(id);
    if (existing !== undefined) {
      return existing;
    }
    const label = `@${this.labels.size + 1}`;
    this.labels.set(id, label);
    return label;
  }
}
